JEST: Quiet Jest tests output on expected failures

mctylr-gh opened the pull request:
This patch quiets the console.error output on expected failures that are suppose to occur during `npm test`. It mocks and restores the console.error function.

For `UserInterface.test.ts` it is limited to only tests that are expected to fail.

In case of `Save.test.ts` is written as a single test, so console.error is muted for the entire test.

mctylr-gh committed:
Quiet Jest tests output on expected failures
This patch quiets the console.error output on expected failures
that are suppose to occur during `npm test`. It mocks and restores
the console.error function.
For `UserInterface.test.ts` it is limited to only tests that are
expected to fail.
In case of `Save.test.ts` is written as a single test, so console.error
is muted for the entire test.

catloversg commented:
Mocking functions like this works, but I suggest using `jest.spyOn` and calling `mockRestore`.

Another problem is `test\jest\Save.test.ts`. The error messages shown in those tests are not false-positive ones. There are real problems with that file. I'll fix them in #2333 later.

mctylr-gh committed:
JEST: Follow-up #2332, use spyOn & expect.toHaveBeenCalled

to improve the quieting of console output by using
jest.spyOn().mockImplementation() and managing the mocked functions
(can use jest.restoreAllMocks()), including adding a test to see that
the code being tested does call console.error during expected failures.

mctylr-gh committed:
JEST: Follow-up #2332 restore test/jest/Save.test.ts

to original behaviour. Those errors be addressed in #2333.

mctylr-gh committed:
JEST: Followup #2332 use mockRestore rather than

`jest.restoreAllMocks()` for better granuality of tests & mocks.

mctylr-gh commented:
I think that takes care of your concerns, both using `jest.spyOn` / `mockRestore` and undoing *my* changes to `test/jest/Save.test.ts`.
